bpo-30814: Fixed a race condition when import a submodule from a package.

serhiy-storchaka opened the pull request.

serhiy-storchaka committed:
bpo-30814: Fixed a race condition when import a submodule from a package.

ncoghlan commented on Lib/importlib/_bootstrap.py:
```diff
@@ -956,9 +956,19 @@ def _find_and_load_unlocked(name, import_):
-    with _ModuleLockManager(name):
-        return _find_and_load_unlocked(name, import_)
+    """Find and load the module."""
+    _imp.acquire_lock()
```
This consolidation seems a little confusing to me, but I like the idea of simplifying the C code by moving the `ModuleNotFoundError` generation into Python.

How about adding a dedicated `_raise_for_halted_import(name, import_):` function that just handles the `module is None` case?

ncoghlan commented on Python/import.c:
```diff
@@ -1532,18 +1532,7 @@ PyImport_ImportModuleLevelObject(PyObject *name, PyObject *globals,
-        goto error;
-    }
-    else if (mod != NULL) {
+    if (mod != NULL && mod != Py_None) {
```
With the separate helper function, we'd keep the dedicated `mod == Py_None` branch, but it would be simplified to just calling the new `_raise_for_halted_import` helper rather than constructing the exception directly in C.

ncoghlan commented on Lib/importlib/_bootstrap.py:
```diff
@@ -956,9 +956,19 @@ def _find_and_load_unlocked(name, import_):
+    module = sys.modules[name]
+    if module is None:
+        _imp.release_lock()
+        message = ('import of {} halted; '
```
This would call `_raise_for_halted_import(name, import_)`.

ncoghlan reviewed (changes requested):
The general approach looks good to me, but I think a dedicated helper function will be clearer when handling the `mod == Py_None` case in C, rather than relying on the second `module is None` check in `_find_and_load`.

serhiy-storchaka commented:
Calling Python method from the C code is not much simpler than generating an exception.

The `module is None` case isn't worth optimization. This isn't common case and isn't performance critical case since raising and catching exception is slow. Only the case of already imported module is worth optimization by implementing in C. It was my mistake that I implemented the generating an exception on C in 133138a284be1985ebd9ec9014f1306b9a425d98, but my patch didn't touched Python code at all.

Note that the new `_find_and_load()` is just a second half of old `_gcd_import()`. All these checks already were implemented in Python, I just moved the code from one Python function to the other.

If you prefer I can withdraw changes in `import.c` and create separate PR for simplifying C code (move locking and generating an exception to Python code) in master only.

I also can implement the second check in C and keep Python code unchanged. But I think that it is better to implement performancy non-critical parts in Python.

ncoghlan commented:
@serhiy-storchaka I think it's specifically the paired negations in `mod != NULL && mod != Py_None` that I find awkward in the C code, so switching the sense of that check to be `mod == NULL || mod == Py_None` (and then swapping the if branches accordingly) would resolve the readability problem without having to change the Python code.

serhiy-storchaka commented:
Do you want to swap also other branches for `fromlist != NULL && fromlist != Py_None` and `!has_from`?

ncoghlan commented:
@serhiy-storchaka I switched my review, as coming back to it after a couple of hours, I now suspect my initial confusion was specifically due to reviewing the *diff*, rather than reading the updated code standalone.

That realisation came mainly from comparing it to the following `fromlist` check that you mentioned (which I didn't find hard to read at all, but which also didn't have a diff confusing the matter).

serhiy-storchaka commented:
Thank you for your review @ncoghlan.
